rasa/core/training/loading.py

```diff
@@ -0,0 +1,146 @@
+import logging
+import os
+from pathlib import Path, PurePath
+from typing import Text, Optional, Dict, List, Union, TYPE_CHECKING
+
+import rasa.utils.io as io_utils
+from rasa.core.domain import Domain
+from rasa.core.interpreter import NaturalLanguageInterpreter, RegexInterpreter
+from rasa.data import YAML_FILE_EXTENSIONS, MARKDOWN_FILE_EXTENSION
+
+from rasa.core.training.structures import StoryStep
+from rasa.core.training.story_reader.story_reader import StoryReader
+from rasa.core.training.story_reader.markdown_story_reader import MarkdownStoryReader
+from rasa.core.training.story_reader.yaml_story_reader import YAMLStoryReader
+
+logger = logging.getLogger(__name__)
+
+
+def _get_reader(
+    filename: Text,
+    domain: Domain,
+    interpreter: NaturalLanguageInterpreter = RegexInterpreter(),
+    template_variables: Optional[Dict] = None,
+    use_e2e: bool = False,
+) -> StoryReader:
+
+    reader = None
+
+    if filename.endswith(MARKDOWN_FILE_EXTENSION):
+        reader = MarkdownStoryReader(
+            interpreter, domain, template_variables, use_e2e, filename
+        )
+    elif PurePath(filename).suffix in YAML_FILE_EXTENSIONS:
+        reader = YAMLStoryReader(
+            interpreter, domain, template_variables, use_e2e, filename
+        )
+    else:
+        # This is a use case for uploading the story over REST API.
+        # The source file has a random name.
+        reader = _guess_reader(
+            filename, domain, interpreter, template_variables, use_e2e
+        )
+
+    if not reader:
+        raise ValueError(
+            f"Failed to find a reader class for the story file `{filename}`. "
+            f"Supported formats are {MARKDOWN_FILE_EXTENSION}, {YAML_FILE_EXTENSIONS}."
+        )
```

Comment:
optional: should `guess_reader` throw instead of throwing here? Then we wouldn't need to check for `if not`